src/transformers/models/deepseek_v3/modular_deepseek_v3.py

```diff
@@ -0,0 +1,434 @@
+            query_states,
+            key_states,
+            value_states,
+            attention_mask,
+            dropout=0.0 if not self.training else self.attention_dropout,
+            scaling=self.scaling,
+            **kwargs,
+        )
+
+        if self.config._attn_implementation == "flash_attention_2" and self.qk_head_dim != self.v_head_dim:
+            attn_output = attn_output[:, :, :, : self.v_head_dim]
+
+        attn_output = attn_output.reshape(batch_size, seq_length, -1).contiguous()
+        attn_output = self.o_proj(attn_output)
+        return attn_output, attn_weights
+
+
+class DeepseekV3DecoderLayer(nn.Module):
+    def __init__(self, config: DeepseekV3Config, layer_idx: int):
+        super().__init__()
+        self.hidden_size = config.hidden_size
+
+        self.self_attn = DeepseekV3Attention(config=config, layer_idx=layer_idx)
+
+        if layer_idx >= config.first_k_dense_replace:
+            self.mlp = DeepseekV3MoE(config)
+        else:
+            self.mlp = DeepseekV3MLP(config)
+
+        self.input_layernorm = DeepseekV3RMSNorm(config.hidden_size, eps=config.rms_norm_eps)
+        self.post_attention_layernorm = DeepseekV3RMSNorm(config.hidden_size, eps=config.rms_norm_eps)
+
+    def forward(
+        self,
+        hidden_states: torch.Tensor,
+        attention_mask: Optional[torch.Tensor] = None,
+        position_ids: Optional[torch.LongTensor] = None,
+        past_key_value: Optional[Cache] = None,
+        output_attentions: Optional[bool] = False,
+        use_cache: Optional[bool] = False,
+        cache_position: Optional[torch.LongTensor] = None,
+        position_embeddings: Optional[Tuple[torch.Tensor, torch.Tensor]] = None,  # necessary, but kept here for BC
+        **kwargs: Unpack[FlashAttentionKwargs],
+    ) -> Tuple[torch.FloatTensor, Optional[Tuple[torch.FloatTensor, torch.FloatTensor]]]:
+        residual = hidden_states
+
+        hidden_states = self.input_layernorm(hidden_states)
+
+        # Self Attention
+        hidden_states, self_attn_weights = self.self_attn(
+            hidden_states=hidden_states,
+            attention_mask=attention_mask,
+            position_ids=position_ids,
+            past_key_value=past_key_value,
+            output_attentions=output_attentions,
+            use_cache=use_cache,
+            cache_position=cache_position,
+            position_embeddings=position_embeddings,
+            **kwargs,
+        )
+        hidden_states = residual + hidden_states
+
+        # Fully Connected
+        residual = hidden_states
+        hidden_states = self.post_attention_layernorm(hidden_states)
+        hidden_states = self.mlp(hidden_states)
+        hidden_states = residual + hidden_states
+
+        outputs = (hidden_states,)
+        if output_attentions:
+            outputs += (self_attn_weights,)
+
+        return outputs
+
+
+class DeepseekV3PreTrainedModel(LlamaPreTrainedModel):
+    pass
+
+
+class DeepseekV3Model(LlamaModel):
```

Comment:
OK, so IF `modeling_deepseek_v3.py` is indeed created from `modular_deepseek_v3.py` (and I am not sure about this), then this. here will not work at all, right? This would simply be the `LlamaModel`.

Reply:
One would at least have to copy the `__init__` and make sure that `DeepseekV3DecoderLayer` is used. But we also need to use `DeepseekV3RotaryEmbedding`, etc.

Reply:
I am still learning. Can you point me to where `modeling` is created from `modular`? My impression is one can change them independently, and there is some automatic check for differences.

I also understand the tole of the `# Copied from ...` comments, one can even specify some transformation rules at the end. But this is not the case for `DeepseekV3Model`. @ArthurZucker  ?

Reply:
OK, and even if this really works, so that I get the classes called `DeepseekXYZ` in `modeling_deepseek_v3.py` automatically from `modular_deepseek_v3.py`, by taking the corresponding code for `LlamaXYZ`, and then vanilla replace "Llama" by "DeepseekV3" everywhere (and I doubt that),

we still need to copy `DeepseekV3RotaryEmbedding` from modeling to modular, because that code genuinely has changed.

Reply:
>One would at least have to copy the __init__ and make sure that DeepseekV3DecoderLayer is used. But we also need to use DeepseekV3RotaryEmbedding, etc. 

that is valid, this is what users should expect, we still need some tests to make sure that we raise an error when the layers being used are not changed as it should be just inheritance!